### PR TITLE
Add unit test cases and make pylint happy

### PR DESCRIPTION
**Pull request by sgandhi1311**
# Description of change

- Added unit test cases for retry logic, exception handling, and sync_endpoint functionality.
- Update the function name `sync` to `sync_stream` to avoid name conflict.

# Manual QA steps
 - Ensured that the tests covered various scenarios, including edge cases.

 
# Risks
 - No Risks
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_outreach/__init__.py [hunk 1]
```diff
@@ -30,7 +30,7 @@ def check_auth(client):
             path='stages',
             endpoint='stages')
     except:
-        raise Exception('Error testing Outreach authentication')
+        raise Exception('Error testing Outreach authentication') from None
```
Comment (RushiT0122) [inline tap_outreach/__init__.py:33]:
```suggestion
    except <ExceptionName> as ex:
        raise Exception('Error testing Outreach authentication') from ex
```

File: tap_outreach/sync.py [hunk 2]
```diff
@@ -281,12 +281,7 @@ def sync_endpoint(client, config, catalog, state, start_date, stream, mdata):
                     filter_field)] = '{}..inf'.format(paginate_datetime)
                 query_params['sort'] = filter_field
 
-        LOGGER.info('{} - Syncing data since {} - page: {}, limit: {}, offset: {}'.format(
-            stream.tap_stream_id,
-            last_datetime,
-            page,
-            count,
-            offset))
+        LOGGER.info(f"{stream.tap_stream_id} - Syncing data since {last_datetime} - page: {page}, limit: {count}, offset: {offset}")
```
Comment (RushiT0122) [inline tap_outreach/sync.py:284]: Line is too long. Also f-string ideally should not be used inside loggers.

Reference: https://discuss.python.org/t/safer-logging-methods-for-f-strings-and-new-style-formatting/13802

File: tap_outreach/sync.py [hunk 3]
```diff
@@ -327,14 +322,14 @@ def update_current_stream(state, stream_name=None):
     singer.write_state(state)
 
 
-def sync(client, config, catalog, state, start_date):
+def sync_stream(client, config, catalog, state, start_date):
```
Comment (RushiT0122) [inline tap_outreach/sync.py:325]: What is the main intension of renaming this method? We use same convention in other taps as well, so IMO we don't need this change unless I am missing something.

File: tests/unittests/test_exception_handling.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+                "mock_filter_field",
+                "mock_fks",
+            )
+        self.assertEqual(e.exception.args[0], 'Only `data` or `links` expected in relationships')
+
+    def test_conflict_data_id(self):
+        """call `process_records` function and by passing the `mock_id_records` in
+        the param.
+        We expect the exception to be raised -
+        `Only `data` or `links` expected in relationships`
```
Comment (RushiT0122) [inline tests/unittests/test_exception_handling.py:61]: This description doesn't look in-line with assertion.

File: tests/unittests/test_exception_handling.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                "mock_filter_field",
+                mock_fks
+            )
+        self.assertEqual(e.exception.args[0], 'null or `id` field expected for `data` relationship')
+
+    def test_conflict_prospects_attributes(self):
+        """call `process_records` function and by passing the `mock_id_records` in
+        the param.
+        We expect the exception to be raised -
+        `Only `data` or `links` expected in relationships`
```
Comment (RushiT0122) [inline tests/unittests/test_exception_handling.py:81]: This description doesn't look in-line with assertion.

File: tests/unittests/test_retry_logic.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+class Mockresponse(Response):
+    def __init__(self, status_code):
+        super().__init__()
+        self.status_code = status_code
+        self.headers = {"x-ratelimit-reset": time.time()}
+    
+    def raise_for_status(self):
+        pass
+    
+class TestRetryLogic(unittest.TestCase):
```
Comment (RushiT0122) [inline tests/unittests/test_retry_logic.py:16]: Add an unittest for non retriable http error code as well.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]
Requested few changes in-line.

File: tap_outreach/sync.py (hunk 3 above)
Comment (sgandhi1311) [reply tap_outreach/sync.py]: referred sync function in __init__.py as _sync to avoid the confusion.

File: tests/unittests/test_retry_logic.py (hunk 6 above)
Comment (sgandhi1311) [reply tests/unittests/test_retry_logic.py:16]: done

File: tests/unittests/test_exception_handling.py (hunk 5 above)
Comment (sgandhi1311) [reply tests/unittests/test_exception_handling.py]: modified

File: tests/unittests/test_exception_handling.py (hunk 4 above)
Comment (sgandhi1311) [reply tests/unittests/test_exception_handling.py]: modified

File: tests/unittests/test_exception_handling.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
+import unittest
+from tap_outreach.sync import process_records
+
+class MockSchema:
```
[the remark this answers is not in the document]
Comment (sgandhi1311) [reply tests/unittests/test_exception_handling.py]: done

**Review by RushiT0122 — APPROVED** [review]
